### PR TITLE
[libra_vm] Refactor libra_vm implementation

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation

Refactored libra_vm into three major components.
1. `LibraVMImpl`, which defines how move functions are being invoked by the execution/validation process.
2. `LibraVM`, which implements the `VMExecutor` api. 
3. `LibraVMValidator`, which implements thr `VMValidator` api.

I got rid of the AccessPathCache for now in the LibraVM due to the borrow_checker issue and we will need to look into that a bit further. The major change in the execution flow is that I got rid of the `VerifiedPayload` and created separate flow for executing the script and executing the module, and a common entry for a failed transaction cleanup.

I intentionally leave `LibraVM` name unchanged just to avoid major renaming in a number of code locations. This can be done in a separate PR.

There were a number of direct move function calls in `LibraVM` such as calling into block_prologue and execute_script body. We might want to get rid of those as well.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

`cargo test`

### CODE REVIEW

**Comment by runtian-zhou** [comment]
Also need some name suggestions here as well for both file names and struct names :)

**Comment by dariorussi** [comment]
> I looked through all of this pretty carefully, and it looks OK to me. I only found one minor thing to suggest. I'll approve, but please make sure @dariorussi is also happy with it.

I am fine with this too and @runtian-zhou and I talked about this today.
I think @runtian-zhou is planning on rewriting this, so it did not need an "approve" :) 
but I agree this looked fine to me too and a good refactor and improvement!

File: language/libra-vm/src/libra_transaction_validator.rs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+    pub fn init_with_config(version: LibraVersion, on_chain_config: VMConfig) -> Self {
+        LibraVMValidator(LibraVMImpl::init_with_config(version, on_chain_config))
+    }
+
+    pub fn load_configs<S: StateView>(&mut self, state: &S) {
+        self.0.load_configs(state)
+    }
+
+    pub fn verify_transaction_impl(
```
Comment (dariorussi) [inline language/libra-vm/src/libra_transaction_validator.rs:47]: does this have to be public?

File: language/libra-vm/src/libra_vm.rs [hunk 2]
```diff
@@ -249,491 +230,41 @@ impl LibraVM {
...
-    fn verify_script(
-        &self,
-        remote_cache: &StateViewCache,
-        script: &Script,
-        txn_data: TransactionMetadata,
-        account_currency_symbol: &IdentStr,
-    ) -> Result<VerifiedTransactionPayload, VMStatus> {
-        let mut cost_strategy = CostStrategy::system(self.get_gas_schedule()?, GasUnits::new(0));
-        let mut session = self.move_vm.new_session(remote_cache);
+    pub fn is_allowed_script(&self, script: &Script) -> Result<(), VMStatus> {
```
Comment (dariorussi) [inline language/libra-vm/src/libra_vm.rs:233]: should most of the functions here by `pub(crate)`

File: language/libra-vm/src/libra_vm.rs [hunk 3]
```diff
@@ -249,491 +230,41 @@ impl LibraVM {
...
-            events,
-            0,
-            TransactionStatus::Keep(VMStatus::executed()),
-        ))
     }
 
     /// Run the prologue of a transaction by calling into `PROLOGUE_NAME` function stored
     /// in the `ACCOUNT_MODULE` on chain.
-    fn run_prologue<R: RemoteCache>(
+    pub fn run_prologue<R: RemoteCache>(
```
Comment (dariorussi) [inline language/libra-vm/src/libra_vm.rs:267]: `pub(crate)`?

**Comment by vgao1996** [comment]
@dariorussi we talked offline, I had a look at the latest version and it looked good to me!

**Pull request closed without merging**